enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+been applied correctly. This step is potentially repeated multiple times by the
+same technician for multiple clusters using the same USB stick or copies of it.
+
+Note that the upgrade bundle should not be specific for a particular cluster,
+only for the OpenShift architecture and version. Technicians should be able to
+use that package for any cluster with that architecture and version.
+
+### API Extensions
+
+There are no new object kinds introduced by this enhancement, but new fields
+will be added to existing `ClusterVersion` and `ContainerRuntimeConfig` objects.
+Those fields are described in detail in the implementation details section below,
+this is a summary of those new `CluserVersion` fields:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  desiredUpdate:
+
+    # Indicates if the upgrade should be performed immediately or if it should
+    # be held for a later time. The default is `false` which means that the
+    # upgrade will be performed immediately (once the images are pre-loaded, if
+    # so requested).
+    hold: true|false
+
+    bundle:
+      # Indicates if use of a bundle is enabled. The default is `true`. If
+      # set to `false` then the `desiredUpdate.image` or `desiredUpdate.version`
+      # field will need to be populated. The cluster version operator will then
+      # pre-load the release images from the registry servers configured in the
+      # cluster and will not use or require a bundle.
+      enabled: true|false
+
+      # Indicates if the cluster version operator should monitor devices
+      # to automatically detect when an upgrade bundle is available. The
+      # defaut is `false`.
+      monitor: true|false
+
+      # Indicates the location of the upgrade bundle when `monitor` is `false`.
+      file: /root/m-bundle.tar # Can be a file.
+      file: /dev/sdb # Or a device.
+
+      # This is an optional digest of the bundle. If set then the cluster version
+      # operator will ensure that the bundle is used only if it matches this
+      # digest. This is intended to prevent use of a wrong USB stick, or use of
+      # a perfectly fine USB stick in the wrong cluster.
+      digest: sha256:....
+
+status:
+  desired:
+    bundle:
+      # Location of the upgrade bundle. This is populated by the monitor, if
+      # enabled, or copied from the `spec.desiredUpdate.bundle.file` field if
+      # monitoring is disabled.
+      file: /root/my-bunde.tar # Can be a file.
+      file: /dev/sdb # Or a device.
+
+      # Node where the bundle is available. This is optional. By default the
+      # cluster version operator will look for the bundle in all the noes, but
+      # if this is used it will look only in the given node.
+      node: node2
+
+      # Contains the current status of the bundle processing in each node
+      # of the cluster. Keys are the node names.
+      nodes:
```

Comment:
How many nodes do these clusters typically have as a max? Should we have a separate CR for each node to avoid bloating the ClusterVersion resource?

Reply:
These are usually small clusters, but we shouldn't rely on that.  Can we just summarize it here?  And if we summarize it here, do we need the detailed information in many other CRs?  If that information is valuable, maybe it can be added as an annotation on the node resource.

Reply:
In the details below I am suggesting to keep here only the basic (and small) information: the boolean flags that control the progress. I am also suggesting to store the large pieces of that (the upgrade metadata, including the list of images) in a separate configmap for each node. That could well be a CR.

Actually in the [prototype](https://github.com/jhernand/upgrade-tool) this information is stored as node annotations, but I think it is too large for that (22 KiB approximately).

Reply:
There will be contention, but not a lot. Take into account that this will only be updated when the bundle is completely extracted and later when it is completely loaded. That will be two updates times the number of nodes. Do you think we should include a new CR for that?